### PR TITLE
Add deterministic ECDSA sig gen. Fix corner cases for add point.

### PR DESCRIPTION
**Pull request by SparkiDev**
In verify, two points are added and they may either be the same point
(different z) or result in infinity.
If they are the same point then a double needs to be performed instead.
Detection of infinity added.
Calculation of wc_ecc_sig_size fixed to handle wehn there are more bits
in the order than in the modulus.

### CODE REVIEW

File: wolfssl/wolfcrypt/ecc.h [hunk 1]
```diff
@@ -391,6 +391,10 @@ struct ecc_key {
     ecc_context_t ctx;
 #endif
 
+#ifdef WOLFSSL_ECDSA_SET_K
+    mp_int sign_k;
```
Comment (toddouska) [inline wolfssl/wolfcrypt/ecc.h]: What about a public key for verify?  Would a pointer be better here to minimize resource use?

Comment (SparkiDev) [reply wolfssl/wolfcrypt/ecc.h]: This is not the private key but the random k value used in signing.
Enabling setting of K to allow known answer tests when signing with ECDSA.

Changed to a pointer as it is a temporary item.

Comment (toddouska) [reply wolfssl/wolfcrypt/ecc.h]: Sure, but ecc_key can be used in the context of a public key for verify only.  And then the user would have to pay for the resource of sign_k.

File: wolfcrypt/src/ecc.c [hunk 2]
```diff
@@ -4939,8 +4964,25 @@ int wc_ecc_sign_hash_ex(const byte* in, word32 inlen, WC_RNG* rng,
...
                     break;
                }
-               err = wc_ecc_make_key_ex(rng, key->dp->size, pubkey,
+       #ifdef WOLFSSL_ECDSA_SET_K
+               if (key->sign_k != NULL) {
+                   if (loop_check > 1) {
+                      err = RNG_FAILURE_E;
+                      break;
+                   }
+                   mp_copy(key->sign_k, &pubkey->k);
```
Comment (toddouska) [inline wolfcrypt/src/ecc.c]: This call can fail in normal math mode.

File: wolfcrypt/src/ecc.c [hunk 3]
```diff
@@ -5347,8 +5426,24 @@ int ecc_mul2add(ecc_point* A, mp_int* kA,
...
+                    /* When all zero then should have done an add */
+                    if (mp_iszero(C->x) && mp_iszero(C->y)) {
+                        err = ecc_projective_dbl_point(precomp[nA + (nB<<2)], C,
+                                                       a, modulus, mp);
+                        if (err != MP_OKAY)
+                            break;
+                    }
+                    /* When only Z zero then result is infinity */
+                    else {
+                        mp_set(C->x, 0);
```
Comment (toddouska) [inline wolfcrypt/src/ecc.c]: These can fail in normal math mode.  Need to check returns.

File: wolfcrypt/src/ecc.c [hunk 4]
```diff
@@ -5861,38 +5956,60 @@ int wc_ecc_verify_hash_ex(mp_int *r, mp_int *s, const byte* hash,
...
+                                                              curve->prime, mp);
+            if (err == MP_OKAY && mp_iszero(mG->z)) {
+                /* When all zero then should have done an add */
+                if (mp_iszero(mG->x) && mp_iszero(mG->y)) {
+                    err = ecc_projective_dbl_point(mQ, mG, curve->Af,
+                                                              curve->prime, mp);
+                }
+                /* When only Z zero then result is infinity */
+                else {
+                    mp_set(mG->x, 0);
```
Comment (toddouska) [inline wolfcrypt/src/ecc.c]: check return.

File: wolfssl/wolfcrypt/ecc.h (hunk 1 above)
Comment (SparkiDev) [reply wolfssl/wolfcrypt/ecc.h]: sign_k is now a pointer in last commit.

File: wolfcrypt/src/ecc.c [hunk 5]
```diff
@@ -8414,25 +8534,41 @@ static int accel_fp_mul(int idx, mp_int* k, ecc_point *R, mp_int* a,
...
+                 /* When all zero then should have done an add */
+                 if (mp_iszero(R->x) && mp_iszero(R->y)) {
+                     if ((err = ecc_projective_dbl_point(fp_cache[idx].LUT[z],
+                                               R, a, modulus, mp)) != MP_OKAY) {
+                         break;
+                     }
+                 }
+                 /* When only Z zero then result is infinity */
+                 else {
+                    mp_set(R->x, 0);
```
Comment (toddouska) [inline wolfcrypt/src/ecc.c]: check return

File: wolfcrypt/src/ecc.c [hunk 6]
```diff
@@ -8635,21 +8771,56 @@ static int accel_fp_mul2add(int idx1, int idx2,
...
+                    if (mp_iszero(R->x) && mp_iszero(R->y)) {
+                        if ((err = ecc_projective_dbl_point(
+                                                  fp_cache[idx1].LUT[zA], R,
+                                                  a, modulus, mp)) != MP_OKAY) {
+                            break;
+                        }
+                    }
+                    /* When only Z zero then result is infinity */
+                    else {
+                       mp_set(R->x, 0);
```
Comment (toddouska) [inline wolfcrypt/src/ecc.c]: check return

File: wolfcrypt/src/ecc.c [hunk 7]
```diff
@@ -8635,21 +8771,56 @@ static int accel_fp_mul2add(int idx1, int idx2,
...
+                    if (mp_iszero(R->x) && mp_iszero(R->y)) {
+                        if ((err = ecc_projective_dbl_point(
+                                                  fp_cache[idx2].LUT[zB], R,
+                                                  a, modulus, mp)) != MP_OKAY) {
+                            break;
+                        }
+                    }
+                    /* When only Z zero then result is infinity */
+                    else {
+                       mp_set(R->x, 0);
```
Comment (toddouska) [inline wolfcrypt/src/ecc.c]: check return

File: wolfcrypt/src/ecc.c [hunk 8]
```diff
@@ -8658,14 +8829,31 @@ static int accel_fp_mul2add(int idx1, int idx2,
...
+                       if (mp_iszero(R->x) && mp_iszero(R->y)) {
+                           if ((err = ecc_projective_dbl_point(
+                                                  fp_cache[idx2].LUT[zB], R,
+                                                  a, modulus, mp)) != MP_OKAY) {
+                               break;
+                           }
+                       }
+                       /* When only Z zero then result is infinity */
+                       else {
+                          mp_set(R->x, 0);
```
Comment (toddouska) [inline wolfcrypt/src/ecc.c]: check return

File: wolfcrypt/src/ecc.c (hunk 2 above)
Comment (SparkiDev) [reply wolfcrypt/src/ecc.c]: fixed

File: wolfcrypt/src/ecc.c (hunk 3 above)
Comment (SparkiDev) [reply wolfcrypt/src/ecc.c]: fixed

File: wolfcrypt/src/ecc.c (hunk 4 above)
Comment (SparkiDev) [reply wolfcrypt/src/ecc.c]: fixed

File: wolfcrypt/src/ecc.c (hunk 5 above)
Comment (SparkiDev) [reply wolfcrypt/src/ecc.c]: fixed

File: wolfcrypt/src/ecc.c (hunk 6 above)
Comment (SparkiDev) [reply wolfcrypt/src/ecc.c]: fixed

File: wolfcrypt/src/ecc.c (hunk 7 above)
Comment (SparkiDev) [reply wolfcrypt/src/ecc.c]: fixed

File: wolfcrypt/src/ecc.c (hunk 8 above)
Comment (SparkiDev) [reply wolfcrypt/src/ecc.c]: fixed

**Pull request merged**